prac_07/project_management.py

```diff
@@ -0,0 +1,241 @@
+    return filename
+
+
+def display_projects(projects):
+    """Displays incomplete projects and then complete projects sorted into """
+    if projects:
+        print("Incomplete projects: ")
+        # print each project in projects sorted by priority if the project is incomplete
+        for project in sorted([project for project in projects if not project.is_complete()]):
+            print(f"\t{project}")
+        print("Completed projects: ")
+        # print each project in projects sorted by priority if the project is complete
+        for project in sorted([project for project in projects if project.is_complete()]):
+            print(f"\t{project}")
+    else:
+        print("No projects to display.")
+
+
+def filter_projects_by_date(projects, selected_filter_date):
+    """Display projects started after date"""
+    if projects:
+        # print each project in projects sorted by the date if the project start date is after date
+        for project in sorted([project for project in projects if project.start_date > selected_filter_date],
+                              key=attrgetter("start_date")):
+            print(f"\t{project}")
+    else:
+        print("No projects to display.")
+
+
+def get_valid_date(prompt):
+    """Get a valid date."""
+    is_valid_date = False
+    while not is_valid_date:
+        try:
+            # Extract the date from a user input like 11/12/13
+            date = datetime.datetime.strptime(input(prompt), "%d/%m/%Y").date()
+            is_valid_date = True
+        except ValueError:
+            print("Invalid date.  Date must be in format dd/mm/yyyy")
+    return date  # Error checking prevents variable being unassigned
+
+
+def add_new_project(projects):
+    """Add a new project to projects."""
+    name = get_valid_name()
+    start_date = get_valid_date("Start date (dd/mm/yyyy): ")
+    priority = get_valid_number(HIGHEST_PRIORITY, LOWEST_PRIORITY, "Priority")
+    cost_estimate = get_valid_cost_estimate(0)
+    completion_percentage = get_valid_number(0, 100, "Percentage completion")
+    projects.append(Project(name, start_date, priority, cost_estimate, completion_percentage))
+
+
+def get_valid_name():
+    """Get a non-empty string."""
+    input_string = input("Name: ")
+    while input_string == "":
+        print("Name cannot be blank")
+        input_string = input("Name: ")
+    return input_string
+
+
+def get_valid_number(minimum, maximum, number_name):
+    """Get a valid number between minimum and maximum inclusive."""
+    is_valid_input = False
+    while not is_valid_input:
+        try:
+            number = int(input(f"{number_name}: "))
+            if minimum <= number <= maximum:
+                is_valid_input = True
+            else:
+                print(f"{number_name} must be > {minimum} and < {maximum}")
+        except ValueError:
+            print("Invalid input.")
+    return number  # Error checking prevents variable being unassigned
+
+
```

Comment:
Overall this code looks very good. I'm just wondering if you can consolidate the get_valid_number and get_valid_cost_estimate into one general purpose input validation function?  Otherwise, everything looks great!

Reply:
Cheers for the review Jarrod 😃 .